### PR TITLE
Issue #978: changes in the formula TF

### PR DESCRIPTION
**Pull request by ermakovpetr**
nltk/nltk/text.py

``` python
def tf(self, term, text, method=None):
    """ The frequency of the term in text. """
    return text.count(term) / len(text)
```

http://en.wikipedia.org/wiki/Tf–idf
Real TF is only the number of occurrences of words in text

And there is also a variation of the "normalization":
1. Boolean "frequencies": tf(t,d) = 1 if t occurs in d and 0 otherwise;
2. Logarithmically scaled frequency: tf(t,d) = 1 + log f(t,d), or zero if f(t, d) is zero;
3. Augmented frequency, to prevent a bias towards longer documents, e.g. raw frequency divided by the maximum raw frequency of any term in the document:
5cc18acd4dbd9be636047fc2a7a10105
4. Normalization of the number of all words
afcade8c6e040203a292c8ca784fa3f3

Source: wiki, Introduction to information retrieval, ...

But before that code, I have not seen anywhere else the normalization of the length of the text.

### CODE REVIEW

File: nltk/text.py [hunk 1]
```diff
@@ -552,11 +552,11 @@ def __init__(self, source, name=None):
         Text.__init__(self, LazyConcatenation(source))
         self._idf_cache = {}
 
-    def tf(self, term, text, method=None):
+    def tf(self, term, text):
         """ The frequency of the term in text. """
-        return text.count(term) / len(text)
```
Comment (ermakovpetr) [inline nltk/text.py:557]: For example: for backward compatibility, I can do this:

``` python
    def tf(self, term, text, canonical=False):
        """ The frequency of the term in text. """
        return text.count(term) / 1 if canonical else len(text)
```

**Comment by dimazest** [comment]
Btw, it might be useful to look at TfidfTransformer [1] and use it and maybe TfidfVectorizer.

[1] http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfTransformer.html#sklearn.feature_extraction.text.TfidfTransformer 

[2] http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html#sklearn.feature_extraction.text.TfidfVectorizer

**Comment by dimazest** [comment]
I like the canonical flag, though maybe rewriting it as

``` python
    def tf(self, term, text, canonical=False):
        """ The frequency of the term in text. """
        if canonical:
                return text.count(term)
        else:
                return text.count(term) / len(text)
```

because it looks less cryptic.

**Comment by ermakovpetr** [comment]
@dimazest I fear for users who expect the previous (is not entirely true) behavior

**Comment by ermakovpetr** [comment]
@dimazest I know TfidfTransformer, but I would like to add the canonical version in my favorite library =)

**Comment by kmike** [comment]
@ermakovpetr a good catch; +1 to using something more common, and +1 to even changing the default. By the way, scikit-learn tf*idf also follows none of wikipedia definitions.

Normalization by the length of text is not totally novel - it is mentioned e.g. here:  http://nlp.stanford.edu/IR-book/html/htmledition/document-and-query-weighting-schemes-1.html - see "normalization" column - though they use len^a, a<1, and in NLTK a==1.

I'm not sure I like 'canonical' flag. Using names from http://nlp.stanford.edu/IR-book/html/htmledition/document-and-query-weighting-schemes-1.html may be a better way to fix it.

**Comment by ermakovpetr** [comment]
@kmike `scikit-learn tf*idf also follows none of wikipedia definitions` 
yes, but (!!!) in Russian wiki tf-idf look like this:
afcade8c6e040203a292c8ca784fa3f3
but I have never seen this method in the non-Russian sources.
In **scikit-learn** tf-idf work like in English wiki (and like different books on English)
Now I want add different normalization in sklearn tf-idf =)

**Comment by stevenbird** [comment]
@larsmans do you have any advice here please?

**Comment by larsmans** [comment]
What NLTK has is apparently [one](https://en.wikipedia.org/w/index.php?title=Tf%E2%80%93idf&oldid=426828805#Mathematical_details) of [several](https://en.wikipedia.org/w/index.php?title=Tf%E2%80%93idf&oldid=508007586#Mathematical_details) non-standard tf-idf definitions that Wikipedia used to have, with no footnote to indicate where it came from. scikit-learn originally had yet another one of these, which was so broken that the weights came out negative. We fixed it to follow _Introduction to IR_, then added an interpolation that made it work better for clustering but has had users confused ever since (scikit-learn/scikit-learn#4391).

The usual definition of tf is just raw term frequency. The [Lucene version](https://lucene.apache.org/core/3_6_0/api/core/org/apache/lucene/search/Similarity.html) takes that square root of that and lots of variants exist. Perhaps the version that NLTK has can be found in [_Modern Information Retrieval_](http://people.ischool.berkeley.edu/~hearst/irbook/); I'm afraid I can't access my copy right now.

**Comment by stevenbird** [comment]
Thanks @larsmans.

**Comment by stevenbird** [comment]
Here's the relevant page from _Modern Information Retrieval_:

image

I didn't see a definition of tf per se.

**Comment by larsmans** [comment]
I think _f_ is supposed to be tf. Check the definition of tf-idf. (IIRC, _MIR_ uses different tf-idf formulas for the query and the documents. You probably want the one for the documents.)

**Comment by stevenbird** [comment]
I propose that we support all definitions of _tf_ listed in the Stanford IR book, with an optional `method` flag, defaulting to raw frequency (`method='natural'`), and drop the existing one (since it's not even listed in the English Wikipedia entry for tf.idf). And likewise for _idf_.

**Comment by stevenbird** [comment]
@ermakovpetr – would you have time to add support for other definitions of tf please?

**Comment by stevenbird** [comment]
Hi @ermakovpetr – just wondering if you would have time to expand the support for tf please.

**Comment by ermakovpetr** [comment]
Hi, @stevenbird.  I have time for this (on next week). I commit in this pull request

**Comment by stevenbird** [comment]
Thanks @ermakovpetr

**Comment by stevenbird** [comment]
I've merged part of this PR; the unimplemented part is documented in #978.

**Comment by ermakovpetr** [comment]
Other push a new branch? (and make a new issue for this)
